share/share.go

```diff
@@ -40,6 +43,31 @@ func GetData(s Share) []byte {
 	return s[NamespaceSize:]
 }
 
+// ShareWithProof contains data with corresponding Merkle Proof
+type ShareWithProof struct { //nolint: revive
+	// Share is a full data including namespace
+	Share
+	// Proof is a Merkle Proof of current share
+	Proof *nmt.Proof
+	// Axis is a type of axis against which the share proof is computed
+	Axis rsmt2d.Axis
+}
+
+// Validate validates inclusion of the share under the given root CID.
+func (s *ShareWithProof) Validate(rootHash []byte, shrIdx, axisIdx, edsSize int) bool {
+	isParity := shrIdx >= edsSize/2 || axisIdx >= edsSize/2
+	namespace := ParitySharesNamespace
+	if !isParity {
+		namespace = GetNamespace(s.Share)
+	}
+	return s.Proof.VerifyInclusion(
+		sha256.New(), // TODO(@Wondertan): This should be defined somewhere globally
```

Comment:
Don't know how old this comment is but some hasher has caused us issues in the past when we tried this, because it couldn't be shared across goroutines. I would assume its the same case here since the hasher can be written to from any goroutine at any time, so I would be careful with this